File: py/vcf/filter.py
from abc import ABCMeta, abstractmethod
from enum import Enum, IntEnum, auto
from typing import Dict, Iterable, Iterator, List, TextIO, Tuple, Union

from vcf.stream import VCFData, VCFDataAction, VCFStream


class VCFFilter(metaclass=ABCMeta):
    def pipe(self, stream: VCFStream) -> VCFStream:
        return _StreamFromFilter(stream, self)
    
    def filter_metadata(self, metadata: List[Tuple[str, str]]) -> List[Tuple[str, str]]:
        return metadata
    
    def filter_individuals(self, individuals: List[str]) -> List[str]:
        return individuals
    
    def filter_item(self, item: VCFData) -> Union[VCFDataAction, VCFData]:
        return item


class _StreamFromFilter(VCFStream):
    def __init__(self, inner: VCFStream, vcf_filter: VCFFilter) -> None:
        self.inner = inner
        self.filter = vcf_filter
        super().__init__()
    
    def start(self) -> None:
        self.metadata = self.filter.filter_metadata(self.inner.metadata)
        self.individuals = self.filter.filter_individuals(self.inner.individuals)
    
    def __next__(self) -> VCFData:
        while True:
            item = next(self.inner)
            action = self.filter.filter_item(item)
            
            if isinstance(action, VCFData):
                return action
            
            if action == VCFDataAction.STOP:
                raise StopIteration


class IndividualsFilter(VCFFilter):
    def __init__(self, individuals: Iterable[str], strict: bool = False) -> None:
        self.individuals = set(individuals)
        self.indices: List[int] = []

        self.strict = strict
    
    def filter_individuals(self, individuals: List[str]) -> List[str]:
        # For each individual in the VCF file, determine whether we are
        # interested in it and keep it only if so
        filtered_individuals = []

        for idx, individual in enumerate(individuals):
            if individual in self.individuals:
                self.indices.append(idx)
                filtered_individuals.append(individual)

        # Also make sure that all individuals of interest have been found
        if self.strict and set(filtered_individuals) != self.individuals:
            bad_individuals = [i for i in self.individuals if i not in filtered_individuals]
            raise Exception(
                f'Cannot find the following individuals in this file: '
                f'{bad_individuals!r}'
            )

        return filtered_individuals

    def filter_item(self, item: VCFData) -> VCFData:
        filtered_fields = [item.data_fields[i] for i in self.indices]
        return VCFData(item.info_fields, filtered_fields)


class PolymorphismFilter(VCFFilter):
    def __init__(self, polymorphisms: Iterable[str]) -> None:
        self.polymorphisms = set(polymorphisms)
        self.count = 0

    def filter_item(self, item: VCFData) -> Union[VCFDataAction, VCFData]:
        if self.count == len(self.polymorphisms):
            return VCFDataAction.STOP
        elif item.get_identifier() not in self.polymorphisms:
            return VCFDataAction.IGNORE
        else:
            self.count += 1
            return item

File: rs/src/vcf/filter.rs
use std::collections::HashSet;
use vcf::stream::{DataStream, IndividualsReader, MetadataReader, VCFData};

pub enum DataAction {
    Ignore,
    Stop,
    Data(VCFData),
}

pub trait Filter {
    fn filter_metadata(&mut self, metadata: Vec<(String, String)>) -> Vec<(String, String)> {
        metadata
    }

    fn filter_individuals(&mut self, individuals: Vec<String>) -> Vec<String> {
        individuals
    }

    fn filter_item(&mut self, item: VCFData) -> DataAction {
        DataAction::Data(item)
    }
}

pub struct Pipe<S: MetadataReader>(S);

impl<S: MetadataReader> Pipe<S> {
    pub fn new(inner: S) -> Pipe<S> {
        Pipe(inner)
    }

    pub fn pipe<F: Filter>(self, filter: F) -> Pipe<impl MetadataReader> {
        Pipe(MetadataReaderFromFilter {
            inner: self.0,
            filter,
        })
    }
    
    pub fn stream(self) -> S {
        self.0
    }
}

struct MetadataReaderFromFilter<S: MetadataReader, F: Filter> {
    inner: S,
    filter: F,
}

impl<S: MetadataReader, F: Filter> MetadataReader for MetadataReaderFromFilter<S, F> {
    type Next = IndividualsReaderFromFilter<S::Next, F>;

    fn read_metadata(
        mut self,
        map: &mut Vec<(String, String)>,
    ) -> Result<Self::Next, &'static str> {
        let mut inner_map = Vec::new();
        let next = self.inner.read_metadata(&mut inner_map)?;
        map.extend(self.filter.filter_metadata(inner_map));

        Ok(IndividualsReaderFromFilter {
            inner: next,
            filter: self.filter,
        })
    }
}

struct IndividualsReaderFromFilter<S: IndividualsReader, F: Filter> {
    inner: S,
    filter: F,
}

impl<S: IndividualsReader, F: Filter> IndividualsReader for IndividualsReaderFromFilter<S, F> {
    type Next = DataStreamFromFilter<S::Next, F>;

    fn read_individuals(mut self, list: &mut Vec<String>) -> Result<Self::Next, &'static str> {
        let mut inner_list = Vec::new();
        let next = self.inner.read_individuals(&mut inner_list)?;
        list.extend(self.filter.filter_individuals(inner_list));

        Ok(DataStreamFromFilter {
            inner: next,
            filter: self.filter,
        })
    }
}

struct DataStreamFromFilter<S: DataStream, F: Filter> {
    inner: S,
    filter: F,
}

impl<S: DataStream, F: Filter> Iterator for DataStreamFromFilter<S, F> {
    type Item = VCFData;

    fn next(&mut self) -> Option<VCFData> {
        loop {
            match self.inner.next().map(|data| self.filter.filter_item(data))? {
                DataAction::Ignore => continue,
                DataAction::Stop => return None,
                DataAction::Data(data) => return Some(data),
            }
        }
    }
}

impl<S: DataStream, F: Filter> DataStream for DataStreamFromFilter<S, F> {}

pub struct IndividualsFilter {
    individuals: HashSet<String>,
    indices: Vec<usize>,
}

impl IndividualsFilter {
    pub fn new<I, Istr>(individuals: I) -> Self
    where
        I: IntoIterator<Item = Istr>,
        Istr: Into<String>,
    {
        IndividualsFilter {
            individuals: individuals.into_iter().map(|s| s.into()).collect(),
            indices: Vec::new(),
        }
    }
}

impl Filter for IndividualsFilter {
    fn filter_individuals(&mut self, individuals: Vec<String>) -> Vec<String> {
        // For each individual in the VCF file, determine whether we are
        // interested in it and keep it only if so
        let mut filtered_individuals = Vec::new();

        for (idx, individual) in individuals.iter().enumerate() {
            if self.individuals.contains(individual) {
                self.indices.push(idx);
                filtered_individuals.push(individual.to_string())
            }
        }

        filtered_individuals
    }

    fn filter_item(&mut self, item: VCFData) -> DataAction {
        DataAction::Data(item.keep(self.indices.iter()))
        // let (mut line, mut tabs) = create_info_line(&item);

        // let (data_line, more_tabs) = create_data_line(&item, line.len(), &self.indices);
        // line.push_str(&data_line);
        // tabs.extend(more_tabs);

        // let new_item = VCFData::new(line, tabs);
        // VCFDataAction::Data(new_item)
    }
}

pub struct PolymorphismFilter {
    polymorphisms: HashSet<String>,
    count: usize,
}

impl PolymorphismFilter {
    pub fn new<I, Pstr>(polymorphisms: I) -> Self
    where
        I: IntoIterator<Item = Pstr>,
        Pstr: Into<String>,
    {
        PolymorphismFilter {
            polymorphisms: polymorphisms.into_iter().map(|s| s.into()).collect(),
            count: 0,
        }
    }
}

impl Filter for PolymorphismFilter {
    fn filter_item(&mut self, item: VCFData) -> DataAction {
        if self.count == self.polymorphisms.len() {
            DataAction::Stop
        } else if !self.polymorphisms.contains(item.identifier()) {
            DataAction::Ignore
        } else {
            self.count += 1;
            DataAction::Data(item)
        }
    }
}
